src/iosrv/sched.c

```diff
@@ -26,73 +26,953 @@
+	void (*func)(void *), void *arg, ABT_thread ult)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_pool_info	*spi;
+	struct sched_request	*req;
+	int			 rc;
+
+	spi = cur_pool_info(info, attr->sra_pool_id);
+	if (spi == NULL) {
+		D_ERROR("XS(%d): get pool info "DF_UUID" failed.\n",
+			dx->dx_xs_id, DP_UUID(attr->sra_pool_id));
+		return NULL;
+	}
+
+	if (d_list_empty(&info->si_idle_list)) {
+		rc = prealloc_requests(info, SCHED_PREALLOC_BATCH_CNT);
+		if (rc)
+			return NULL;
+	}
+
+	req = d_list_entry(info->si_idle_list.next, struct sched_request,
+			   sr_link);
+	d_list_del_init(&req->sr_link);
+
+	req->sr_attr	= *attr;
+	req->sr_func	= func;
+	req->sr_arg	= arg;
+	req->sr_ult	= ult;
+	req->sr_abort	= 0;
+	req->sr_pool_info = spi;
+
+	return req;
+}
+
+static void
+req_put(struct dss_xstream *dx, struct sched_request *req)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+
+	D_ASSERT(d_list_empty(&req->sr_link));
+	/* Don't put into idle list when the request is tracked by caller */
+	if (req->sr_ult == ABT_THREAD_NULL)
+		d_list_add_tail(&req->sr_link, &info->si_idle_list);
+}
+
+static int
+req_kickoff_internal(struct dss_xstream *dx, struct sched_req_attr *attr,
+		     void (*func)(void *), void *arg)
+{
+	ABT_pool	abt_pool;
+	int		rc;
+
+	D_ASSERT(attr && func && arg);
+	switch (attr->sra_type) {
+	case SCHED_REQ_IO:
+		abt_pool = dx->dx_pools[DSS_POOL_IO];
+		break;
+	case SCHED_REQ_GC:
+		abt_pool = dx->dx_pools[DSS_POOL_GC];
+		break;
+	case SCHED_REQ_MIGRATE:
+		abt_pool = dx->dx_pools[DSS_POOL_REBUILD];
+		break;
+	default:
+		D_ASSERTF(0, "Invalid req type: %u\n", attr->sra_type);
+		break;
+	}
+
+	rc = ABT_thread_create(abt_pool, func, arg, ABT_THREAD_ATTR_NULL, NULL);
+
+	return dss_abterr2der(rc);
+}
+
+static int
+req_kickoff(struct dss_xstream *dx, struct sched_request *req)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_pool_info	*spi = req->sr_pool_info;
+	struct sched_req_info	*sri;
+	int			 rc;
+
+	if (req->sr_ult != ABT_THREAD_NULL) {
+		rc = ABT_thread_resume(req->sr_ult);
+		rc = dss_abterr2der(rc);
+	} else {
+		rc = req_kickoff_internal(dx, &req->sr_attr, req->sr_func,
+					  req->sr_arg);
+	}
+
+	D_ASSERT(spi != NULL);
+	D_ASSERT(req->sr_attr.sra_type < SCHED_REQ_MAX);
+	sri = &spi->spi_req_array[req->sr_attr.sra_type];
+
+	D_ASSERT(sri->sri_req_cnt > 0);
+	sri->sri_req_cnt--;
+	D_ASSERT(info->si_req_cnt > 0);
+	info->si_req_cnt--;
+
+	d_list_del_init(&req->sr_link);
+	req_put(dx, req);
+
+	return rc;
+}
+
+#define SCHED_SPACE_AGE_MAX	5000	/* 5000 msecs */
+
+static int
+check_space_pressure(struct dss_xstream *dx, struct sched_pool_info *spi,
+		     bool purge)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct vos_pool_space	 vps = { 0 };
+	uint64_t		 scm_left, nvme_left;
+	int			 rc;
+
+	D_ASSERT(spi->spi_space_ts <= info->si_cur_ts);
+	/* Use cached space presure info */
+	if ((spi->spi_space_ts + SCHED_SPACE_AGE_MAX) > info->si_cur_ts)
```

Comment:
not sure if check pool space is very cheap, or maybe check it very 5S is too frequent.

Reply:
This is a light weight check, only one SCM read and few DRAM reads. I measured latency & iops with 1 second MAX_AGE, didn't observe non-negligible performance drop.

Actually I'm going to shrink this value to 2 seconds to get more up-to-date space information.